clients/drcachesim/tools/filter/func_id_filter.h

```diff
@@ -0,0 +1,141 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#ifndef _FUNC_ID_FILTER_H_
+#define _FUNC_ID_FILTER_H_ 1
+
+#include "record_filter.h"
+#include "trace_entry.h"
+#include "utils.h"
+
+#include <cstring>
+#include <unordered_set>
+#include <vector>
+
+namespace dynamorio {
+namespace drmemtrace {
+
+/* This filter takes a list of function IDs for which it preserves the related
+ * TRACE_MARKER_TYPE_FUNC_[ID | ARG | RETVAL | RETADDR] markers. It removes all the other
+ * TRACE_MARKER_TYPE_FUNC_ markers related to function IDs that are not in the list.
+ */
+class func_id_filter_t : public record_filter_t::record_filter_func_t {
+public:
+    func_id_filter_t(std::vector<uint64_t> keep_func_ids_list)
+    {
+        for (auto keep_func_id : keep_func_ids_list) {
+            keep_func_ids_set_.insert(keep_func_id);
+        }
+    }
+
+    void *
+    parallel_shard_init(memtrace_stream_t *shard_stream,
+                        bool partial_trace_filter) override
+    {
+        per_shard_t *per_shard = new per_shard_t;
+        per_shard->output_func_markers = false;
+        return per_shard;
+    }
+
+    bool
+    parallel_shard_filter(
+        trace_entry_t &entry, void *shard_data,
+        record_filter_t::record_filter_info_t &record_filter_info) override
+    {
+        /* Get per_shard private data.
+         */
+        per_shard_t *per_shard = reinterpret_cast<per_shard_t *>(shard_data);
+
+        trace_type_t entry_type = static_cast<trace_type_t>(entry.type);
+        if (entry_type == TRACE_TYPE_MARKER) {
```

Comment:
Invert this to reduce indentation: return if it's not a marker, with the if body now at top level.

Reply:
Done.